src/MOLFiniteDifference/MOL_discretization.jl

```diff
@@ -165,18 +169,48 @@ function SciMLBase.symbolic_discretize(pdesys::ModelingToolkit.PDESystem,discret
-        #                        *(~~a..., ~~b..., dot(reverse_weights(i,j),depvarsdisc[k][central_neighbor_idxs(i,j)[1:2]])),
-        #                        *(~~a..., ~~b..., dot(forward_weights(i,j),depvarsdisc[k][central_neighbor_idxs(i,j)[2:3]]))))
-        #                        for j in 1:nspace, k in 1:length(depvars)]
-    
-        rules = vcat(vec(central_deriv_rules),valrules)
+        forward_weights(II,j) = DiffEqOperators.calculate_weights(discretization.upwind_order, 0.0, grid[j][[II[j],II[j]+1]])
+        reverse_weights(II,j) = DiffEqOperators.calculate_weights(discretization.upwind_order, 0.0, grid[j][[II[j]-1,II[j]]])
+        # upwinding_rules = [@rule(*(~~a,$(Differential(nottime[j]))(u),~~b) => IfElse.ifelse(*(~~a..., ~~b...,)>0,
+        #                         *(~~a..., ~~b..., dot(reverse_weights(II,j),depvars[k][central_neighbor_idxs(II,j)[1:2]])),
+        #                         *(~~a..., ~~b..., dot(forward_weights(II,j),depvars[k][central_neighbor_idxs(II,j)[2:3]]))))
+        #                         for j in 1:nspace, k in 1:length(pdesys.depvars)]
+
+        ## Discretization of non-linear laplacian. 
+        # d/dx( a du/dx ) ~ (a(x+1/2) * (u[i+1] - u[i]) - a(x-1/2) * (u[i] - u[i-1]) / dx^2
+        b1(II, j, k) = dot(reverse_weights(II, j), depvarsdisc[k][central_neighbor_idxs(II, j)[1:2]]) / dxs[j]
+        b2(II, j, k) = dot(forward_weights(II, j), depvarsdisc[k][central_neighbor_idxs(II, j)[2:3]]) / dxs[j]
+        # TODO: improve interpolation of g(x) = u(x) for calculating u(x+-dx/2)
+        g(II, j, k, l) = sum([depvarsdisc[k][central_neighbor_idxs(II, j)][s] for s in (l == 1 ? [2,3] : [1,2])]) / 2.
+        # iv_mid returns middle space values. E.g. x(i-1/2) or y(i+1/2).
+        iv_mid(II, j, l) = (grid[j][II[j]] + grid[j][II[j]+l]) / 2.0 
+        # Dependent variable rules
+        r_mid_dep(II, j, k, l) = [depvars[k] => g(II, j, k, l) for k in 1:length(depvars)]
+        # Independent variable rules
+        r_mid_indep(II, j, l) = [nottime[j] => iv_mid(II, j, l) for j in 1:length(nottime)]
+        # Replacement rules: new approach
+        rules = [@rule ($(Differential(iv))(*(~~a, $(Differential(iv))(dv), ~~b))) =>
+                 dot([Num(substitute(substitute(*(~~a..., ~~b...), r_mid_dep(II, j, k, -1)), r_mid_indep(II, j, -1))),
+                      Num(substitute(substitute(*(~~a..., ~~b...), r_mid_dep(II, j, k, 1)), r_mid_indep(II, j, 1)))],
+                     [-b1(II, j, k), b2(II, j, k)])
+                 for (j, iv) in enumerate(nottime) for (k, dv) in enumerate(depvars)]
+        rhs_arg = (SymbolicUtils.operation(eq.rhs) == +) ? SymbolicUtils.arguments(eq.rhs) : [eq.rhs]
+        lhs_arg = (SymbolicUtils.operation(eq.lhs) == +) ? SymbolicUtils.arguments(eq.lhs) : [eq.lhs]
+        nonlinlap_rules = []
+        for t in vcat(lhs_arg,rhs_arg)
+            for r in rules
+                if r(t) !== nothing
```

Comment:
?

Reply:
what exactly is this doing?

Reply:
This code allows the non linear laplacian rules (`rules`) to be used as an argument of the function `substitute`, in an analogous way to what is done with the central derivative rules (see the code below). The first `for` iterates over the terms of the equation. The second `for` iterates over the non-linear laplacian rules (thus, a version of the rules that can not be used as an argument of the function `substitute`). If the current rule can be applied to the current term, a new rule in the correct format is generated.
Another option is not to use the `substitute` function to apply the nonlinear laplacian rules, but if we want to be consistent with what was done with the central derivative rules, In think that we have to make a transformation. Suggestions are welcome :)

Reply:
No but what is `r(t) !== nothing` actually doing, and is that public API?

Reply:
`r(t)` is applying the rule `r` to the term `t`, if the rule does not match with the term it returns `nothing`, otherwise it returns the transformed term. When we use `@rule`  (see code above) we need to work with the rule as it were a function, therefore we cannot directly use `substitute` as it is used with the central derivative rules. 

It is in the public API ("https://symbolicutils.juliasymbolics.org/api/"):

> `rule LHS => RHS`
> Creates a `Rule` object. A rule object is callable, and takes an expression and rewrites it if it matches the LHS pattern to the RHS pattern, returns nothing otherwise.

Reply:
`r(t) !== nothing` is asking if the current rule can be applied to the current term.

Reply:
It should probably just be `!=` but that's not a huge deal.

Reply:
Great, I will fix it in the next PR :)

Reply:
that's my bad, VScode had a blue underline under `!=` which disappeared when I changed to `!==`